simplicity-sys/src/dag.rs

```diff
@@ -0,0 +1,120 @@
+// Simplicity Bindings
+// Written in 2022 by
+//   Christian Lewe <clewe@blockstream.com>
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+use crate::bitstream::Bitstream;
+use crate::bitstring::Bitstring;
+use crate::error::Error;
+use libc::{c_void, size_t};
+use std::ops::Index;
+
+/// Kind of Simplicity node.
+///
+/// Corresponds to `tag_t` in `depend/simplicity/dag.h`.
+#[repr(C)]
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub enum Tag {
+    COMP,
+    CASE,
+    ASSERTL,
+    ASSERTR,
+    PAIR,
+    DISCONNECT,
+    INJL,
+    INJR,
+    TAKE,
+    DROP,
+    IDEN,
+    UNIT,
+    HIDDEN,
+    WITNESS,
+    JET,
+}
+
+/// (First node of) a Simplicity DAG (typed or untyped).
+///
+/// Corresponds to `dag_node` in `depend/simplicity/dag.h`.
+#[repr(C)]
+pub struct DagNode {
+    jet: *const c_void,
+    cmr: [u32; 8],
+    // Assume largest union member
+    union_aux_types: [size_t; 2],
+    // Assume largest union member
+    union_indices_children_witness: Bitstring,
+    tag: Tag,
+}
+
+extern "C" {
+    fn decodeMallocDag(
+        dag: *mut *mut DagNode,
+        combinator_counters: *const c_void,
+        stream: *mut Bitstream,
+    ) -> i32;
+}
+
+impl DagNode {
+    /// Decode a Simplicity DAG from bytes.
+    /// Returns the DAG and its length on success.
+    pub fn decode(bytes: &[u8]) -> Result<(&mut DagNode, usize), Error> {
+        let mut bitstream = Bitstream::from(bytes);
+        let mut dag: *mut DagNode = std::ptr::null_mut();
```

Comment:
Do we have to explicitly free the DAG in the destructor of `DagNode`? [In the C code, the DAG is freed after use.](https://github.com/ElementsProject/simplicity/blob/master/C/test.c#L152)

Reply:
Yep, you need to free `dag` in the `Drop` impl of `DagNode`. Good catch.